---
title: "0.Mapping_pipeline"
author: "Enrico"
date: "30 January 2019"
output: html_document
---

In order to calculate demographic models with machine learning methods (step carried by other group = ), high coverage sequencing data for at least 2 individuals per species will be mapped to latest version of the cat reference genome (version 9.0 -> downloaded from ensembl and copied to /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome folder).

```{r, eval=FALSE, engine='bash'}

# Reference Genome needs indexing
samtools faidx /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa

java -jar /opt/picard-tools/picard.jar CreateSequenceDictionary R= /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa O= /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.dict

bwa index /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa

```

Since the samples were already quality checked and trimmed for previous projects, these steps were skipped and I proceded to map directly.

# 0. Path Definition

Define the paths to different locations

```{r, eval=FALSE, engine='bash'}

MacroGenARRAY=($(ls /backup/grupolince/raw_data/MACROGEN/MACROGEN_trimmed/*.fastq.gz | rev | cut -d'/' -f 1 | rev | cut -d '_' -f1 | uniq)) # List of all Lynx lynx sample codes in MACROGEN project

REF=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa # path to cat reference genome

THREADS=10 # No. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!

OUT=/home/ebazzicalupo/CatRef_bams # path to output files

MacroGenPATH=/backup/grupolince/raw_data/MACROGEN/MACROGEN_trimmed/ # path to macrogen fastq files


# BARCODE MACROGEN:
declare -A BARCODEID=(["LC1"]="c_lc_zz_0001" ["LL112"]="c_ll_vl_0112" ["LL146"]="c_ll_ya_0146" ["LL212"]="c_ll_cr_0212" ["LL90"]="c_ll_ki_0090" ["LR1"]="c_lr_zz_0001")

```

## 	Mapping BWA-MEM	

```{r, engine=bash, eval=FALSE}

# Trial with only second (1) array element -> LL112

screen -S CatRef_Mapping.log
script CatRef_Mapping.log

for i in ${MacroGenARRAY[1]}
  do
    echo " - mapping ${i} -"
    bwa mem $REF $MacroGenPATH/${i}_R1_trimmed.fastq.gz $MacroGenPATH/${i}_R2_trimmed.fastq.gz -t $THREADS | samtools view -hbS -@ $THREADS - -o $OUT/${i}.cat_ref.bam
    echo " - sorting ${i} -"
    samtools sort -@ $THREADS $OUT/${i}.cat_ref.bam -o $OUT/${i}.cat_ref.sorted.bam && rm $OUT/${i}.cat_ref.bam
    echo " - done -"
done

```

## 	Add Read Groups and Change BAM names

```{r, engine=bash, eval=FALSE}

# Trial with only second (1) array element -> LL112

# Read Groups

screen -S ReadGroups.log
script ReadGroups.log

for i in ${MacroGenARRAY[1]}
  do
    run=($(echo $i | cut -d"_" -f 1))  #Sacar el run de i
    echo $run
    java -jar /opt/picard-tools/picard.jar AddOrReplaceReadGroups I=$OUT/${i}.cat_ref.sorted.bam O=$OUT/${BARCODEID["${i}"]}_${i}.cat_ref.sorted.rg.bam RGID=${i} RGLB=${BARCODEID["${i}"]}_lib RGPL=Illumina RGPU=${run} RGSM=${BARCODEID["${i}"]} VALIDATION_STRINGENCY=SILENT && rm $OUT/${i}.cat_ref.sorted.bam
done

# Merge BAMs / Change BAM names

SAMPLESLIST=($(echo ${BARCODEID[@]} | tr ' ' '\n' | sort | uniq | grep c_ll_vl_0112))
for sample in "${SAMPLESLIST[@]}"
  do
    echo "${sample}"
    ls $OUT/"${sample}"_*.cat_ref.sorted.rg.bam  > $OUT/"${sample}".bam.list
    echo;echo
    echo `wc -l $OUT/"${sample}".bam.list`;
    lines=`wc -l $OUT/"${sample}".bam.list | cut -f 1 -d " " `
    if [ "$lines" -eq "1" ]; 
      then echo "ONE";
      cp `cat $OUT/"${sample}".bam.list`  $OUT/"${sample}"_sorted.bam;
    elif [ "$lines" -gt "1" ]; 
      then echo "more than ONE";
      samtools merge  -@ $THREADS -r $OUT/"${sample}".bam `cat $OUT/"${sample}".bam.list`;
      samtools sort  -@ $THREADS $OUT/"${sample}".bam -o $OUT/"${sample}"_sorted.bam && rm $OUT/"${sample}".bam;
    fi
    samtools flagstat $OUT/"${sample}"_sorted.bam >  $OUT/"${sample}".stats;
done

```

## 	Remove / Mark Duplicates

```{r, engine=bash, eval=FALSE}

screen -S MDUP.log
script MDUP.log

ARRAY_MERGED_BAM_SAMPLE_NAME=($(ls $OUT/*_sorted.bam | rev | cut -d'/' -f1 | rev | cut -d'_' -f1-4 | uniq))
for i in ${ARRAY_MERGED_BAM_SAMPLE_NAME[@]}
  do
  echo "${i}"
  java -jar /opt/picard-tools/picard.jar MarkDuplicates METRICS_FILE=${i}_rmdup.txt I=$OUT/${i}_sorted.bam O=$OUT/${i}_sorted_rmdup.bam MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800
  rm $OUT/${i}_sorted.bam
  samtools sort $OUT/${i}_sorted_rmdup.bam -@ 10 -o $OUT/${i}_sorted_rmdup_sorted.bam
  rm $OUT/${i}_sorted_rmdup.bam
  samtools index $OUT/${i}_sorted_rmdup_sorted.bam
  samtools flagstat $OUT/${i}_sorted_rmdup_sorted.bam > $OUT/${i}_sorted_rmdup_sorted.stats
done

```

##  Realign

```{r, engine=bash, eval=FALSE}

screen -S realign.log
script realign.log

ARRAY_MERGED_BAM_SAMPLE_NAME=($(ls $OUT/*_sorted.bam | rev | cut -d'/' -f1 | rev | cut -d'_' -f1-4 | uniq))

## Not INCLUDING -fixMisencodedQuals flag because MACROGEN are Illumina 1.9 (quality code doesn't need fixing)

for i in ${ARRAY_MERGED_BAM_SAMPLE_NAME[@]}
  do
    echo "${i}"
    samtools index $OUT/${i}_sorted_rmdup_sorted.bam
    # RealignerTargetCreator
    java -jar /home/tmp/Software/GATK_3.4/GenomeAnalysisTK.jar -T RealignerTargetCreator -nt 10 -R $REF -I $OUT/${i}_sorted_rmdup_sorted.bam -o $OUT/${i}_realignertargetcreator.intervals
    # IndelRealigner
    java -jar /home/tmp/Software/GATK_3.4/GenomeAnalysisTK.jar -T IndelRealigner -R $REF -targetIntervals $OUT/${i}_realignertargetcreator.intervals -I $OUT/${i}_sorted_rmdup_sorted.bam -o $OUT/${i}_sorted_rmdup_sorted_indelrealigner.bam

done

```

##  Base Quality Score Recalibration

```{r, engine=bash, eval=FALSE}

```
