---
title: "0.Mapping_5x_cesga_pipeline"
author: "Enrico"
date: "30 May 2019"
output: html_document
---

With this markdown I explain the steps I took to test and run alignments to the newest cat reference genome of low depth sequencing data from a few populations of both Iberian and Eurasian lynxes.

At first I will map Iberian lynx individuals from Sierra Morena and Eurasian lynx individuals from Kirov, Carpathians and Vladivostok.

As the alignments will be conducted on the CESGA ft2 server, I will first need to test if all the softwares are available and working there. Then I will create SBATCH (the queue system of ft2) scripts that will align the samples.

The softwares required during mapping are (in order of how they are used):

(1) BWA : mem
(2) SAMtools : view, sort, merge
(3) PICARDtools : AddOrReplaceReadGroups, MarkDuplicates
(4) GATK : RealignerTargetCreator, IndelRealigner
