---
title: "1.Calling_pipeline"
author: "Enrico"
date: "9 April 2019"
output: html_document
---

With the BAMs from the 20 high coverage/depth individuals we will perform a variant calling using GATK.
Base quality score recalibration has not been conducted on the BAMs as we don't have a set of variants to mask before recalibration.
We might use the first round of calling for recalibration or not.

## GATK 4.1.0.0

I will follow GATK 4.1.0.0 best practices and do:

  (1) a primary per-sample calling, using the tool HaplotypeCaller, that will generate a GVCF file for each sample (using a GVCF specific flag while running the program);

  (2) GVCF consolidation (GenomicsDBImport) which consists in taking the previously generated GVCFs and creating a Database for the third step to work with, would be the standard best practice advised by GATK. The problem with this program is that it runs only on determined intervals of the genome. Since we are using whole genome data it is best to use the CombineGVCFs tool, which, although slower, doesn't require intervals.

  (3) finally, a joint-calling of variants (GenotypeGVCFs) using the samples pre-called with HaplotypeCaller and combined with CombineGVCFs.

The step of filtering and recalibration of variants will not be conducted with the best practices GATK protocol, as we are missing a reliable set of variants to mask out during this analysis. Again, we might use the first round of calling for recalibration or not.


### Preparing the data

For the final calling the data is already ready, as only the BAM files and the reference genome are needed.
I first would like to try with just one chromosome from two individuals (c_ll_vl_0112 and c_lr_nm_0006) to test how the program runs.
I need to generate a BED file with the coordinates of the chromosome I want to use (A1). This information is stored in the FAI index file of the reference genome.

```
# Create a BED with coordinates from chromosome A1
cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | grep "A1" \
| cut -f1,2 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, 0, $2}' > /home/ebazzicalupo/try_calling/A1_CHR_coordinates.bed

# Subset the BAM to that region only - Tried on c_ll_vl_0112 - run in a screen
samtools view -b -h -L /home/ebazzicalupo/try_calling/A1_CHR_coordinates.bed \
/home/ebazzicalupo/CatRef_bams/c_ll_vl_0112_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.bam \
> /home/ebazzicalupo/try_calling/c_ll_vl_0112_cat_ref_chrA1.bam
# Index the subset sample
samtools index /home/ebazzicalupo/try_calling/c_ll_vl_0112_cat_ref_chrA1.bam


# Subset the BAM to that region only - Also on c_lr_nm_0006 - run in a screen
samtools view -b -h -L /home/ebazzicalupo/try_calling/A1_CHR_coordinates.bed \
/home/ebazzicalupo/CatRef_bams/c_lr_nm_0006_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.bam \
> /home/ebazzicalupo/try_calling/c_lr_nm_0006_cat_ref_chrA1.bam
# Index the subset sample
samtools index /home/ebazzicalupo/try_calling/c_lr_nm_0006_cat_ref_chrA1.bam

```

### Variables and Paths Definition

Here are all the variables I used during the trial version with just the two individuals.

```
# Reference Genome:
REF=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa

# Input Files: - it's the same as the output for the trial version
IN=/home/ebazzicalupo/try_calling

# Output Files:
OUT=/home/ebazzicalupo/try_calling

# A1 chromosome BED file:
A1BED=/home/ebazzicalupo/try_calling/A1_CHR_coordinates.bed

```

### Per-Sample Variant Calling - HaplotypeCaller

This is the first step of the pipeline.
I will run it for the two individuals (c_ll_vl_0112 and c_lr_nm_0006) subsampled for just chromosome A1.

##### A1 chromosome BAM

```
# Per-Sample calling of c_ll_vl_0112

screen -S c_ll_vl_0112_trycall.log
script c_ll_vl_0112_trycall.log

/opt/gatk-4.1.0.0/gatk --java-options "-Xmx4g" HaplotypeCaller  \
   -R $REF \
   -I $IN/c_ll_vl_0112_cat_ref_chrA1.bam \
   -O $OUT/c_ll_vl_0112_cat_ref_chrA1.g.vcf.gz \
   -ERC GVCF

# Per-Sample calling of c_lr_nm_0006

screen -S c_lr_nm_0006_trycall.log
script c_lr_nm_0006_trycall.log

/opt/gatk-4.1.0.0/gatk --java-options "-Xmx4g" HaplotypeCaller  \
  -R $REF \
  -I $IN/c_lr_nm_0006_cat_ref_chrA1.bam \
  -O $OUT/c_lr_nm_0006_cat_ref_chrA1.g.vcf.gz \
  -ERC GVCF

```
This was very SLOW (~30-40 hours for just one chromosome). Talk with Dani or Arturo about what to change. Possible Xmx flag is limiting??

##### No Flags

Too check if it was the -Xmx creating problems, I will try it with no flags

```
screen -S c_ll_vl_0112_noflag_trycall.log
script c_ll_vl_0112_noflag_trycall.log

/opt/gatk-4.1.0.0/gatk HaplotypeCaller  \
   -R $REF \
   -I $IN/c_ll_vl_0112_cat_ref_chrA1.bam \
   -O $OUT/c_ll_vl_0112_cat_ref_chrA1.g.vcf.gz \
   -ERC GVCF

```
This was also slow = ~30h

##### Increased Xmx

I'll try increasing the Xmx flag value to dedicate more memory to GATK and see if performance improves.

```
screen -S c_ll_vl_0112_xmx64G_trycall.log
script c_ll_vl_0112_xmx64G_trycall.log

/opt/gatk-4.1.0.0/gatk --java-options "-Xmx64g" HaplotypeCaller  \
   -R $REF \
   -I $IN/c_ll_vl_0112_cat_ref_chrA1.bam \
   -O $OUT/c_ll_vl_0112_cat_ref_xmx64G_chrA1.g.vcf.gz \
   -ERC GVCF

```
This was also slow = ~30h

##### Increased HMM threads

I'll also try increasing the threads for the pairHMM algorithm. The default value is 4, maybe increasing might improve performance.

```
screen -S c_ll_vl_0112_15threads_trycall.log
script c_ll_vl_0112_15threads_trycall.log

/opt/gatk-4.1.0.0/gatk HaplotypeCaller  \
   -R $REF \
   -I $IN/c_ll_vl_0112_cat_ref_chrA1.bam \
   -O $OUT/c_ll_vl_0112_cat_ref_15threads_chrA1.g.vcf.gz \
   --native-pair-hmm-threads 15 \
   -ERC GVCF

```
This was also slow = ~30h (a bit less than the rest)

##### Whole BAM with -L A1BED

Adjusting the previous parameters hasn't helped improving the performance (except the last one only very slightly, so I included it). I'll try changing the approach by using the BAM file with all of the alignment and giving GATK an archive with the positions of interest (the A1 chromosome BED).

```
screen -S c_ll_vl_0112_A1BED_trycall.log
script c_ll_vl_0112_A1BED_trycall.log

/opt/gatk-4.1.0.0/gatk HaplotypeCaller  \
   -R $REF \
   -I /home/ebazzicalupo/CatRef_bams/c_ll_vl_0112_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.bam \
   -O $OUT/c_ll_vl_0112_cat_ref_A1BED_chrA1.g.vcf.gz \
   --native-pair-hmm-threads 15 \
   -L $A1BED \
   -ERC GVCF

```

### GVCF Combining - CombineGVCFs

This is the second step of the pipeline.

```
screen -S try_CombineGVCFs.log
script try_CombineGVCFs.log

/opt/gatk-4.1.0.0/gatk CombineGVCFs \
   -R $REF \
   --variant $OUT/c_ll_vl_0112_cat_ref_chrA1.g.vcf.gz \
   --variant $OUT/c_lr_nm_0006_cat_ref_chrA1.g.vcf.gz \
   -O $OUT/try_chrA1.g.vcf.gz

```

### Joint-Calling of variants - GenotypeGVCFs

This is the third step of the pipeline.

```
screen -S try_GenotypeGVCFs.log
script try_GenotypeGVCFs.log

/opt/gatk-4.1.0.0/gatk --java-options "-Xmx4g" GenotypeGVCFs \
  -R $REF \
  -V $OUT/try_chrA1.g.vcf.gz \
  -O $OUT/try_chrA1_joint.vcf.gz

```

Unless significant improvements can be made to the speed of the HaplotypeCaller, I don't believe GATK 4.1.0.0 is a good option.

## GATK 3.7

Because of the slowness of GATK 4.1.0.0 HaplotypeCaller, I decided to try version 3.7 for comparison.

### Per-Sample Variant Calling - HaplotypeCaller

This is the first step of the pipeline.
As before, I will run it for the two individuals (c_ll_vl_0112 and c_lr_nm_0006) subsampled for just chromosome A1.

##### A1 chromosome BAM

```
screen -S GATK3.7_c_ll_vl_0112_trycall.log
script GATK3.7_c_ll_vl_0112_trycall.log

java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -jar \
  /opt/GATK-3.7/GenomeAnalysisTK.jar \
  -T HaplotypeCaller \
  -R $REF \
  -I $IN/c_ll_vl_0112_cat_ref_chrA1.bam \
  --emitRefConfidence GVCF \
  -o $OUT/c_ll_vl_0112_cat_ref_GATK3.7_chrA1.g.vcf

```

This is terribly slow! Maybe the problem is with the BAM itself.

##### Whole BAM with -L A1BED

As I've done with GATK 4.1.0.0, I'll try with GATK 3.7 to run the HaplotypeCaller with the entire BAM, giving GATK an archive with the positions of interest (the A1 chromosome BED).

```
screen -S GATK3.7_c_ll_vl_0112_A1BED_trycall.log
script GATK3.7_c_ll_vl_0112_A1BED_trycall.log

java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -jar \
  /opt/GATK-3.7/GenomeAnalysisTK.jar \
  -T HaplotypeCaller \
  -R $REF \
  -I /home/ebazzicalupo/CatRef_bams/c_ll_vl_0112_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.bam \
  -o $OUT/c_ll_vl_0112_cat_ref_A1BED_chrA1.g.vcf \
  -L $A1BED \
  --emitRefConfidence GVCF

```
