---
title: "1.Parallel_Calling_pipeline"
author: "Enrico"
date: "30 April 2019"
output: html_document
---

With the BAMs from the 20 high coverage/depth individuals we will perform a variant calling using GATK.
Base quality score recalibration has not been conducted on the BAMs as we don't have a set of variants to mask before recalibration.
We might use the first round of calling for recalibration or not.

With the trials described in "1.Calling_pipeline.md" I have concluded that a parallelization step is necessary to run the analysis in reasonable time.

In this markdown I explain the trials I have done to generate an executable program that will parallelize the variant calling. I will use GATK version 4.1.0.0

To put it really shortly:

if I have an array of Chromosomes and Samples, I can run, for each chromosome, a command that will open a screen for each sample, executing GATK HaplotypeCaller for that sample. Because of hardware limitations (64 cores in total and a soft limit of 20 cores per user) I can only run one chromosome at the time (20 individuals = 20 cores at the time).

## Variables

```
# Array of Chromosomes
CHRarray=($(cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | cut -f 1 | uniq))

# Array of BED files
BEDarray=($(ls /home/ebazzicalupo/CatGenome_CHR_BEDs/*.bed | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1 | uniq))

# Array of BAM files (samples)
SAMPLEarray=($(ls /home/ebazzicalupo/CatRef_bams/*.bam | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1,2,3,4 | uniq))

```
## Preparing chromosome BED files

Each chromosome's coordinates must be written in a BED file to be given to GATK. Because there are 21 "regular" chromosomes and ~4500 very short chromosomes, I will make a BED for each "regular" one and a single BED for the "rest".

```
# Create a BED with coordinates for each "regular" chromosome

for i in ${CHRarray[@]:0:20}
  do
  echo $i
  cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | grep "$i" \
  | cut -f1,2 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, 0, $2}' > /home/ebazzicalupo/CatGenome_CHR_BEDs/"$i"_CHR_coordinates.bed
done

```

```
# Create a BED with coordinates the "rest" of chromosomes - remove and remake BED first as we are using ">>"

rm /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
touch /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
for i in ${CHRarray[@]:20:4508}
  do
  echo $i
  cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | grep "$i" \
  | cut -f1,2 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, 0, $2}' >> /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
done

```

## Screen for one Chromosomes and every Sample

I want a script that can go through the array of chromosome BED files, and for one of them open a screen for each sample. This way I can just select the array element I prefer (work with one chromosome), or run for all of the chromosome BEDs one by one in a loop (I'm scared of doing this if something goes wrong, so probably I'll just work with one at the time).

In this case it will only echo the sample and bed names inside of the screen, to test if the command works.

```
for bed in ${BEDarray[0]}
  do
    echo ${bed}
    for sample in ${SAMPLEarray[@]}
      do
        screen -dmS "${sample}_${bed}_calling"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "script "${sample}_${bed}_calling".log\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "echo ${sample}_${bed}; exec bash\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "exit\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "exit\n"
    done
done

```
By changing the 0 with any other number in the array you can choose to run the script for the different Chromosomes.

This small will be implemented in script (chrarrayscript).sh, to which you will just have to specify a particular BED array element (with a number) and the script will run (samplearrayscript).sh for that chromosome.

The (samplearrayscript).sh will use the values ${sample} and ${bed} from the (chrarrayscript).sh in GATK (similarly to how you specify the BED array element), to specify which sample and which chromosome HaplotypeCaller has to work on.
