---
title: "1.Parallel_Calling_pipeline"
author: "Enrico"
date: "30 April 2019"
output: html_document
---

With the BAMs from the 20 high coverage/depth individuals we will perform a variant calling using GATK.
Base quality score recalibration has not been conducted on the BAMs as we don't have a set of variants to mask before recalibration.
We might use the first round of calling for recalibration or not.

With the trials described in "1.Calling_pipeline.md" I have concluded that a parallelization step is necessary to run the analysis in reasonable time.

In this markdown I explain the trials I have done to generate an executable program that will parallelize the variant calling. I will use GATK version 4.1.0.0

To put it really shortly:

if I have an array of Chromosomes and Samples, I can run, for each chromosome, a command that will open a screen for each sample, executing GATK HaplotypeCaller for that sample. Because of hardware limitations (64 cores in total and a soft limit of 20 cores per user) I can only run one chromosome at the time (20 individuals = 20 cores at the time).

I SWITCHED TO CESGA FT2 SERVER WHICH HAS MUCH HIGHER COMPUTING POWER. See below!

## Variables

```
# Array of Chromosomes
CHRarray=($(cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | cut -f 1 | uniq))

# Array of BED files
BEDarray=($(ls /home/ebazzicalupo/CatGenome_CHR_BEDs/*.bed | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1 | uniq))

# Array of BAM files (samples)
SAMPLEarray=($(ls /home/ebazzicalupo/CatRef_bams/*.bam | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1,2,3,4 | uniq))

```
## Preparing chromosome BED files

Each chromosome's coordinates must be written in a BED file to be given to GATK. Because there are 21 "regular" chromosomes and ~4500 very short chromosomes, I will make a BED for each "regular" one and a single BED for the "rest".

```
# Create a BED with coordinates for each "regular" chromosome

for i in ${CHRarray[@]:0:20}
  do
  echo $i
  cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | grep "$i" \
  | cut -f1,2 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, 0, $2}' > /home/ebazzicalupo/CatGenome_CHR_BEDs/"$i"_CHR_coordinates.bed
done

```

```
# Create a BED with coordinates the "rest" of chromosomes - remove and remake BED first as we are using ">>"

rm /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
touch /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
for i in ${CHRarray[@]:20:4508}
  do
  echo $i
  cat /home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | grep "$i" \
  | cut -f1,2 | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1, 0, $2}' >> /home/ebazzicalupo/CatGenome_CHR_BEDs/rest_CHR_coordinates.bed
done

```

## Screen for one Chromosomes and every Sample

I want a script that can go through the array of chromosome BED files, and for one of them open a screen for each sample. This way I can just select the array element I prefer (work with one chromosome), or run for all of the chromosome BEDs one by one in a loop (I'm scared of doing this if something goes wrong, so probably I'll just work with one at the time).

In this case it will only echo the sample and bed names inside of the screen, to test if the command works.

```
for bed in ${BEDarray[0]}
  do
    echo ${bed}
    for sample in ${SAMPLEarray[@]}
      do
        screen -dmS "${sample}_${bed}_calling"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "script "${sample}_${bed}_calling".log\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "echo ${sample}_${bed}; exec bash\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "exit\n"
        screen -S "${sample}_${bed}_calling" -p 0 -X stuff "exit\n"
    done
done

```
By changing the 0 with any other number in the array you can choose to run the script for the different Chromosomes.

This small will be implemented in script chr_parallel_caller.sh, to which you will just have to specify a particular BED array element (with a number) and the script will run sample_chr_Haplotypecaller.sh for that chromosome.

The sample_chr_Haplotypecaller.sh will use the values ${sample} and ${bed} from the chr_parallel_caller.sh in GATK (similarly to how you specify the BED array element), to specify which sample and which chromosome HaplotypeCaller has to work on.

## CESGA ft2 SERVER

As the CESGA "finis terrae II" server has much higher computing power (and possibly faster running times) I will conduct the variant calling there.

In order to run a script (execute a job) on the ft2 server, you need to send it to the queue system through the command sbatch. A few parameters can be added to the script in order to personalize your requirements to the system:

(1) Execution time limit : I will put "indefinite" because I don't know how long the analysis will take
(2) Partition : I will use the thinnodes partition which is the standard one for normal workloads
(3) Number of cores : I will give 24 cores as is the maximum amount for one node (need to check!)
(4) Send emails : I will tell the script to send an automatic email to me as the job starts and as it ends

My script will include these parameters and the GATK HaplotypeCaller command.

### Copying the data on ft2 from genomics.ebd

I will pass my BAM files to the ft2 server going through this array one element at the time (because I will need to insert the password manually each time).

```
SAMPLEarray=($(ls /home/ebazzicalupo/CatRef_bams/*.bam | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1,2,3,4 | uniq | grep -E -v "ll_cr_0212|lp_sm_0359|lr_nm_0006"))

# 0 = c_lc_zz_0001
# 1 = c_lc_zz_0003
# 2 = c_ll_ki_0090
# 3 = c_ll_vl_0112
# 4 = c_ll_ya_0146
# 5 = c_lp_do_0153
# 6 = c_lp_do_0173
# 7 = c_lp_do_0443
# 8 = c_lp_sm_0138
# 9 = c_lp_sm_0140
# 10 = c_lp_sm_0185
# 11 = c_lp_sm_0186
# 12 = c_lp_sm_0221
# 13 = c_lp_sm_0298
# 14 = c_lr_fl_0005
# 15 = c_lr_zz_0001
# 16 = h_lp_do_0007

for sample in ${SAMPLEarray[16]}
	do
	echo "copying $sample"
	scp /home/ebazzicalupo/CatRef_bams/"$sample"_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.ba* csebdjg2@ft2.cesga.es:/mnt/lustre/scratch/home/csic/ebd/jg2/test/CatRef_bams
	echo "done"
done

```

### Submitting jobs in ft2

The script will be very straightforward and will launch the calling for every individual (see 1.Callling-executables/cesga_Haplotypecaller.sh). With a loop it will be lauched for every individual separately.

It appears that for default settings the maximum run time is of 4 days. This means that I will have to submit 1 chromosome at the time. So the script should include also the target chromosome. I can probably submit jobs for all the individuals for 2 chromosomes at the time.

I will run this script one BEDarray element at the time (one chromosome) to respect the per-user limits of CESGA.

Job submission numbers will be tracked in CESGA_jobsubmission_IDs.md

```
# Array of BED files
BEDarray=($(ls $LUSTRE/test/CatGenome_CHR_BEDs/*.bed | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1 | uniq))

# 0 = A1
# 1 = A2
# 2 = A3
# 3 = B1
# 4 = B2
# 5 = B3
# 6 = B4
# 7 = C1
# 8 = C2
# 9 = D1
# 10 = D2
# 11 = D3
# 12 = D4
# 13 = E1
# 14 = E2
# 15 = E3
# 16 = F1
# 17 = F2
# 18 = MT
# 19 = rest
# 20 = X

# List of Samples - FOR CHR A1 (BEDarray# = 0) ONLY: without lp_sm_0359 which was already called during the trial ( add :  | grep -v "lp_sm_0359" to the array)
SAMPLEarray=($(ls $LUSTRE/test/CatRef_bams/*.bam | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 1,2,3,4 | uniq))

# Loop of job submission to the queue

for bed in ${BEDarray[5]}
  do
    echo ${bed}
    for sample in ${SAMPLEarray[@]}
      do
        echo "launching $sample"
        sbatch cesga_Haplotypecaller.sh ${sample} ${bed}
    done
done

```
