---
title: "2.Variant_Filtering"
author: "Enrico"
date: "30 April 2019"
output: html_document
---

Next step in the pipeline is to filter unwanted variants from the final VCF file.

This are the criteria I will use to filter variants:

(1) Repetitive/Low mappability regions
(2) Indels + Non-biallelic sites
(3) Lynx genus wide exclusive substitutions from Felis catus
(4) Hard quality filters, as GATK standard practices
(5) Over/Under covered regions (<5% and >95% of depth distribution)
(6) Under-represented, excessively missing variants
(7) Genes?

I will try these filters individually in an interactive node of FT2 using a single chromosome VCF to see if any errors appear (see below for trial commands/scripts - keep in mind that the paths to files might not be well defined and can be inconsistent). After these trials I will write the scripts to apply to the final, whole-genome VCF.

Filters 1 to 4 can be applied to the whole dataset. Filters 5 and 6 need to be calculated at the level of the individual/species. For this reason I will write a first "General-Filters" script, that applies filters 1 to 4.

## (1) Repetitive/Low mappability regions

In order to filter repetitive and low mappability regions I needed to extract the coordinates of those regions in a BED format. The reference genome has a version (marked as rm in the file name = Felis_catus.Felis_catus_9.0.dna_rm.toplevel.fa) with these regions masked with Ns:

a TAGAGAGAGAGATT segment would be TAGNNNNNNNNATT

More information on how these regions were masked can be found in the README file of the reference genome folder available in the ensembl database.

With a python script I found online (at https://www.danielecook.com/generate-a-bedfile-of-masked-ranges-a-fasta-file/) I was able to create the BED file with the masked regions coordinates. You can find the python script in 2.Variant_Filtering-executables/generate_masked_ranges.py

```
# This was done in the $LUSTRE/test/CatGenome_Masked_BEDs/ folder in ft2 where the script
# and the masked reference genome are

./generate_masked_ranges.py Felis_catus.Felis_catus_9.0.dna_rm.toplevel.fa > Masked_Regions.bed

```

Let's try to filter the A1 chromosome for these coordinates.

```
# Initial number of variants:
zcat $LUSTRE/test/CatRef_vcfs/A1_cat_ref.vcf.gz | grep -v "#" | wc -l # 7417329

# Apply the filter with BedTools subtract
zcat $LUSTRE/test/CatRef_vcfs/A1_cat_ref.vcf.gz | bedtools subtract -a - -b $LUSTRE/test/CatGenome_Masked_BEDs/Masked_Regions.bed -header | uniq > $LUSTRE/test/FilterTrials/A1_cat_ref.vcf

# Number of variants after filtering:
grep -v "#" $LUSTRE/test/FilterTrials/A1_cat_ref.vcf | wc -l # 3801893

# Number of variants filtered:
# 7417329 - 3615436 = 3801893

```

## (2) Indels + Non-biallelic sites

I want to filter out INDELs and Non-Biallelic positions, because they represent a kind of data which is much harder to analyze.

I will use the GATK tool SelectVariants which has a two flags specifically for filtering these types of variants.

As before, let's try using this command with our A1 VCF file.

```


```
