---
title: "2.Variant_Filtering"
author: "Enrico"
date: "30 April 2019"
output: html_document
---

Next step in the pipeline is to filter unwanted variants from the final VCF file.

This are the criteria I will use to filter variants:

(1) Repetitive/Low mappability regions
(2) Indels + Non-biallelic sites
(3) Lynx genus wide exclusive substitutions from Felis catus
(4) Hard quality filters, as GATK standard practices
(5) Over/Under covered regions (<5% and >95% of depth distribution)
(6) Under-represented, excessively missing variants
(7) Heterozigosity?
(8) Genes?

## Repetitive/Low mappability regions

In order to filter repetitive and low mappability regions I needed to extract the coordinates of those regions in a BED format. The reference genome has a version (marked as rm in the file name = Felis_catus.Felis_catus_9.0.dna_rm.toplevel.fa) with these regions masked with Ns:

a TAGAGAGAGAGATT segment would be TAGNNNNNNNNATT

More information on how these regions were masked can be found in the README file of the reference genome folder available in the ensembl database.

With a python script I found online (at https://www.danielecook.com/generate-a-bedfile-of-masked-ranges-a-fasta-file/) I was able to create the BED file with the masked regions coordinates. You can find the python script in 2.Variant_Filtering-executables/generate_masked_ranges.py

```
# This was done in the $LUSTRE/test/CatGenome_Masked_BEDs/ folder in ft2 where the script
# and the masked reference genome are

./generate_masked_ranges.py Felis_catus.Felis_catus_9.0.dna_rm.toplevel.fa > Masked_Regions.bed

```
