---
title: "3.Admixture_structure"
author: "Enrico"
date: "5 February 2020"
output: html_document
---

Having successfully called and filtered the variants of our dataset, I will now use different softwares and methods to test for admixture between species.

This markdown explains my use of the software Admixture 1.3.0 to check for structure between species and possibly admixed individuals.

## Preparing the data

I first need to transform my VCF to PLINK format using VCF tools.
```
vcftools --gzvcf WholeGenome_cat_ref.filter7.vcf.gz --plink --out WholeGenome_plink
```
This generates a PLINK format .ped and .map file. Because Felis catus chromosome names are not recognised by PLINK, I need to manually modify the chromosome name column for each SNP in the .map file.

The information about chromosome name is still saved in the second column of the .map file, as the SNP name is assigned as chromosomename:position.

The only chromosome names recognised by PLINK are the human ones. I need to write a script that will change the 0 in the first column (un-recognised chromosome names are encoded as 0s) to 1 if the second column starts with A1, 2 if it starts with A2, and so on. This way I change Cat chromosome names that are letter+number to just a number.

This will be done in two consecutive steps. First I will change the chromosome name for all autosomal SNPs (corresponding human chromosome number 1 to 18). Rest of genome (X chromosome, Mitochondrial genome, and non-chromosomal contigs) will have a different encoding (X and Mit are recognised by PLINK and rest will be discarded and left with a value of 0 in the chromosome column of .map file), so I will copy them as they are.
```
# list of chromosomes in map file (from second column)
CHRarray=($(cut -f2 WholeGenome_plink.map | cut -d':' -f1 | uniq | head -18))

# make sure to start with an empty new .map file
rm WholeGenome_plink_Humchr.map
touch WholeGenome_plink_Humchr.map

# Change to human number for autosomal SNPs
for i in {0..17}
do
let "j=$i+1"
echo $j
grep "${CHRarray[$i]}" ~/ALDER/INPUT/WholeGenome_plink.map | sed "s/0/${j}/" \
>> WholeGenome_plink_Humchr.map
done

# Add non-autosomal ones as they are.
for i in "${CHRarray[@]}"
do
echo "$i" >> chrlist
done
grep -v -f chrlist WholeGenome_plink.map >> WholeGenome_plink_Humchr.map

# Finally, change name to .ped to match the new .map name
mv WholeGenome_plink.ped WholeGenome_plink_Humchr.ped
```
Then I need to transform the PLINK (.ped and .map) to PLINK 1 binary file (.bed , .bim , .fam) and filter out non-autosomal data.
```
plink_1.9 --file WholeGenome_plink_Humchr --chr 1-18 --make-bed --out WholeGenome_plink_Humchr_binary
```
Now that I have my final PLINK binary files, I need to subset my samples, in order to have a uniform sample size across species. To do this I will remove 7 of my 11 Lynx pardinus samples, to be left with 2 canada, 4 eurasian, 4 iberian lynxes and 3 bobcats.
```
# Make list of samples to keep (grep command exclude the listed samples)
grep -vE "c_lp_do_0443|h_lp_do_0007|c_lp_sm_0185|c_lp_sm_0186|c_lp_sm_0221|c_lp_sm_0298|c_lp_sm_0359" WholeGenome_plink_Humchr_binary.fam \
> 4lp_dataset.txt
```
Prune dataset with previously listed samples and for LD, as suggested in Admixture software manual (changed r2 between SNPs to be at least 0.3).
```
# creates a list of snps to keep
plink_1.9 --bfile WholeGenome_plink_Humchr_binary --indep-pairwise 50 10 0.3 --out r2_0.3
# filter the dataset
plink_1.9 --bfile WholeGenome_plink_Humchr_binary --extract r2_0.3.prune.in --keep 4lp_dataset.txt \
--make-bed --out WholeGenome_plink_Humchr_r2_0.3_4lp
```

## Run the Analysis

Run the analysis looping through K = 3 to 10 (K is the number of groups), 10 times each, to have a good amount of repetitions and compare between them.
Bootstrapping is not necessary as it gives confidence intervals for the admixture proportions, in which we are not interested in.

We are not looking at differences in admixture proportions, so I will repeat multiple times to check for alternative scenarios, rather than having a confidence interval for the only scenario inferred in one analysis.
```
for i in 1 2 3 4 5 6 7 8 9 10
  do
   mkdir rep_${i}
   cd rep_${i}
   for k in 3 4 5 6 7 8
     do
      echo "run number $i of admixture for k = $k"
      /opt/admixture_linux-1.3.0/admixture \
      /home/ebazzicalupo/Admixture_structure/input/WholeGenome_plink_Humchr_r2_0.3_4lp.bed \
      $k -j5 --cv | tee log_r2_0.3_4lp_rep${i}_k${k}.out
   done
   cd ..
done
```
This way I have a rep_"n" folder for each repetition, containing the results for K between 3 and 8.
I need to download these on my local computer to paint results in R
```
scp -r ebazzicalupo@genomics-b.ebd.csic.es:/home/ebazzicalupo/Admixture_structure/rep_* ~/admixture_structure/
```
This was done to extract the calculated CV value for each K repetition.
```
# get CVs for each k value in log file
for i in 3 4 5 6 7 8; do cat rep_*/log_r2_0.3_4lp_rep*_k${i}* | grep "CV" > k${i}.cv.log; done
```
The results are painted in R using the R script found in the 3.Admixture-executables/ folder.
