---
title: "3.Admixture_test_Treemix"
author: "Enrico"
date: "23 September 2019"
output: html_document
---

Having successfully called and filtered the variants of our dataset, I will now use different softwares and methods to test for admixture between species.

This markdown explains my use of the software Treemix, version 1.12.

A first round of trials will be run interactively on the genomics-b server, which already has Treemix 1.12 installed. When the dataset will be ready (waiting for Oscar to send VCF with genes filtered out), and I will need to run repetitions of the analysis, the same pipeline tested here will be run on the CESGA ft2 server (MAYBE).

## Preparing Input Files

The first step is to prepare the input files necessary for Treemix 1.12.

### Intergenic region BED

As I want to exclude genes, which might bias Treemix's analysis, I will generate a BED file with coordinates of all the genes, so I can filter out all of the variants that are found within these regions, plus a buffer of 1k bp before and after the gene. I will extract the information from the GFF3 annotation file of the reference genome (Felis catus).

```
REFdir=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome

# Extract gene coordinates from Reference Genome Annotation
grep "ID=gene" $REFdir/Felis_catus.Felis_catus_9.0.97.chr.gff3 | cut -f 1,4,5 | awk 'BEGIN { OFS = "\t" } {
  if($2 >= 1000)
  {
    print $1,$2-=1000,$3+=1000
  }
  else
  {
    print $1,0,$3+=1000
  }
}' > $REFdir/Felis_catus.Felis_catus_9.0.97.GENES_1kbuffer.bed
```

### PerSpecies VCF files

As I need an allele count for each population (species in my case), I will have to first divide my VCF into 4 different VCFs, one for each species. Doing this with GATK will guarantee that the allele count (AC) and allele number (AN) fields will be recalculated.

```
screen -S GATKsppPart
script GATKsppPart.log


# Define an array with the species codes (lp,ll,lc,lr)
speciesARRAY=($(ls ~/CatRef_bams/*.bam | rev | cut -d'/' -f1 | rev | cut -d'_' -f2 | sort -u))

# Define Reference Genome
REF=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa

# Define VCF Directory
VCF=/home/ebazzicalupo/CatRef_vcfs

# create a BAMlist file for each species listing all BAMs of that species
for species in ${speciesARRAY[@]}
  do
  ls ~/CatRef_bams/*.bam | grep _$species > ~/CatRef_bams/${species}.bamlist
done


# for each species, run GATK to select only individuals belonging to that species

for species in ${speciesARRAY[@]}
  do

  samplesARRAY=($(cat ~/CatRef_bams/${species}.bamlist | rev | cut -d'/' -f1 | rev | cut -d'_' -f1-4))

  /opt/gatk-4.1.0.0/gatk SelectVariants \
  -R $REF \
  -V $VCF/WholeGenome_cat_ref.filter7.vcf \
  $(for j in ${samplesARRAY[@]}; do echo "-sn ${j}";done) \
  -O $VCF/${species}.vcf

done
```

### Remove Variants of Gene regions and Create Allele Count file

Next step will be to remove any variants that fall within gene regions, to obtain a supposedly "neutral" dataset of variants. This will be done with the command "subtract" of BEDtools.

```
# Define an array with the species codes (lp,ll,lc,lr)
speciesARRAY=($(ls ~/CatRef_bams/*.bam | rev | cut -d'/' -f1 | rev | cut -d'_' -f2 | sort -u))

# Define reference genome directory
REFdir=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome

for species in ${speciesARRAY[@]}
  do

  # Remove genese from VCF
  bedtools subtract -a ~/CatRef_vcfs/"$species".vcf \
  -b $REFdir/Felis_catus.Felis_catus_9.0.97.GENES_1kbuffer.bed -header \
  > ~/CatRef_vcfs/"$species".NOgenes.vcf

  # Extract allele counts and write file
  grep -v "#" ~/CatRef_vcfs/"$species".NOgenes.vcf | cut -d';' -f1,3 | \
  grep -o -E 'AC=[[:digit:]]{1,3};AN=[[:digit:]]{1,3}' | \
  sed 's/AC=//g' | sed 's/;AN=/,/g' | awk -F',' 'BEGIN { OFS = "," } {print $2-$1,$1}' \
  > ~/Treemix/"$species".allelecounts

done

# Generate final allele count file
echo "lc ll lp lr" > All.allelecounts
paste -d' ' lc.allelecounts ll.allelecounts lp.allelecounts lr.allelecounts >> All.allelecounts

# gZip it!
gzip All.allelecounts
```

## Running Treemix 1.12

Now that our input file is ready we can run Treemix.

To account for the fact that nearby SNPs are not independent, we will group them together in windows of size 1000 SNPs, using the -k flag. This way we account for LD, which might affect the results of our analysis.

Our idea is to model admixture onto two different genus tree topologies. One will be the tree inferred by Treemix. The other will be a previously inferred topology, reconstructed by our collaborator William Murphy in a recent paper. This last topology is based on admixture free areas of the genome, and should be more accurate in reconstructing relationships between species.

The command to impose a particular tree topology is -tf [file]. The file (topology.tree) should contain a Newick format tree, with distances.

Example:
(((lc:0.3,ll:0.3):0.3,lp:0.3):0.3,lr:0.3);

The distance values will be calculated ....

Apart from the different tree topologies, different number of events of migration will also be tested. The way Treemix works, we can then plot the fraction of the variance, to see what is the most likely number of migration events. Because of this, Treemix (with both inferred and predetermined topology) will be run with increasing number of migration events.

```
############################################
# Run Treemix Without pre-defined topology #
############################################

screen -S Treemix_notree
script Treemix_notree.log

cd ~/Treemix

# Run Treemix for sequentially more migration events
for i in $(seq 1 10)
  do
  echo "#############################"
  echo "running treemix with m = ${i}"
  echo "#############################"
  mkdir m_${i}
  treemix -i All.allelecounts.gz -k 1000 -m ${i} -o m_${i}/treemix_out
done


```
