---
title: "3.Admixture_test_Treemix"
author: "Enrico"
date: "23 September 2019"
output: html_document
---

Having successfully called and filtered the variants of our dataset, I will now use different softwares and methods to test for admixture between species.

This markdown explains my use of the software Treemix, version 1.XX (???).

A first round of trials will be run on the genomics-b server, which already has Treemix 1.12 installed. When the dataset will be ready (waiting for Oscar to send VCF with genes filtered out), and I will need to run repetitions of the analysis, the same pipeline tested here will be run on the CESGA ft2 server (see below).

## STEP 1 - preparing input files

The first step is to prepare the input files necessary for Treemix 1.XX.

### Intergenic region BED

As I want to exclude genes, which might bias Treemix's analysis, I will generate a BED file with coordinates of all the genes, so I can filter out all of the variants that are found within these regions, plus a buffer of 1k bp before and after the gene. I will extract the information from the GFF3 annotation file of the reference genome (Felis catus).

```
REFdir=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome

# Extract gene coordinates from Reference Genome Annotation
grep "ID=gene" $REFdir/Felis_catus.Felis_catus_9.0.97.chr.gff3 | cut -f 1,4,5 | awk '$2-=1000,$3+=1000' \
> $REFdir/Felis_catus.Felis_catus_9.0.97.GENES_1kbuffer.bed
```

### PerSpecies VCF files

As I need an allele count for each population (species in my case), I will have to first divide my VCF into 4 different VCFs, one for each species. Doing this with GATK will guarantee that the allele count (AC) and allele number (AN) fields will be recalculated.

```
screen -S GATKsppPart
script GATKsppPart.log


# Define an array with the species codes (lp,ll,lc,lr)
speciesARRAY=($(ls ~/CatRef_bams/*.bam | rev | cut -d'/' -f1 | rev | cut -d'_' -f2 | sort -u))

# Define Reference Genome
REF=/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa

# Define VCF Directory
VCF=/home/ebazzicalupo/CatRef_vcfs

# create a BAMlist file for each species listing all BAMs of that species
for species in ${speciesARRAY[@]}
  do
  ls ~/CatRef_bams/*.bam | grep _$species > ~/CatRef_bams/${species}.bamlist
done


# for each species, run GATK to select only individuals belonging to that species

for species in ${speciesARRAY[@]}
  do

  samplesARRAY=($(cat ~/CatRef_bams/${species}.bamlist | rev | cut -d'/' -f1 | rev | cut -d'_' -f1-4))

  /opt/gatk-4.1.0.0/gatk SelectVariants \
  -R $REF \
  -V $VCF/WholeGenome_cat_ref.filter7.vcf \
  $(for j in ${samplesARRAY[@]}; do echo "-sn ${j}";done) \
  -O $VCF/${species}.vcf

done

```
