---
title: "4.PSMC_analysis"
author: "Enrico"
date: "10 February 2020"
output: html_document
---

I will try running the PSMC program following the tutorial I found at:
https://informatics.fas.harvard.edu/psmc-journal-club-walkthrough.html


## PATH DEFINITION

This will be done on the genomics EBD server. I defined the paths to different folders for easier execution and script writing:
```
BAM_path="/home/ebazzicalupo/CatRef_bams"
REF="/home/GRUPOS/grupolince/reference_genomes/felis_catus_genome/Felis_catus.Felis_catus_9.0.dna.toplevel.fa"
CONSENSUS_path="/home/ebazzicalupo/PSMC/consensus_fq"
PSMCinput_path="/home/ebazzicalupo/PSMC/inputfiles_psmcfa"
OUTPUT_path="/home/ebazzicalupo/PSMC/output_psmc"
```

## 1. GENERATE CONSENSUS SEQUENCES

First step is to generate consensus sequences from BAM files.

You generate a consensus with a pipeline of SAMtools mpileup, BCFtools call and a vcf utility which converts a vcf to fq, which can be converted to the input format for PSMC by a utility of PSMC (see below).
```
# To try it out I will only run it on chromosome F2 (smaller autosomal)
CHR="F2"

# Pipe samtools mpileup, to bcftools call, to vcfutils.pl for final .fq file
samtools mpileup -Q 30 -q 30 -u -v \
-f $REF -r $CHR $BAM_path/c_lr_fl_0005_cat_ref_sorted_rg_rmdup_sorted_indelrealigner.bam |
bcftools call -c |
vcfutils.pl vcf2fq -d 5 -D 34 -Q 30 > $CONSENSUS_path/c_lr_fl_0005_cat_ref.$CHR.fq

# convert the FQ file to the PSMC input format
/opt/psmc/utils/fq2psmcfa $CONSENSUS_path/c_lr_fl_0005_cat_ref.F2.fq > $PSMCinput_path/c_lr_fl_0005_cat_ref.F2.psmcfa
```
An alternative method can be found here:
https://samtools.github.io/bcftools/howtos/consensus-sequence.html


## 2. RUN PSMC

I will run PSMC on the generated consensus sequence using default parameters (they can be adjusted).
```
# Run PSMC on the file
psmc -p "4+25*2+4+6" -o $OUTPUT_path/c_lr_fl_0005_cat_ref.F2.psmc $PSMCinput_path/c_lr_fl_0005_cat_ref.F2.psmcfa

# Now to draw the results (-u = mutation rate; -g = generation time)
/opt/psmc/utils/psmc_plot.pl -u 1.6e-08 -g 5 $OUTPUT_path/c_lr_fl_0005_cat_ref.F2.plot $OUTPUT_path/c_lr_fl_0005_cat_ref.F2.psmc
```

## SCRIPT to run PSMC for all samples

Using the above trials, I wrote a script to run the same steps on the Whole Genome Sequences of all the samples of the project.

The script can be found at:
4.PSMC-executables/PSMC-executable.sh


## Paint PSMC results

I painted all plots of the different samples individually with a loop:
```
PSMC_out_ARRAY=($(ls $OUTPUT_path))

for i in ${PSMC_out_ARRAY[@]}
	do
	id=($(echo $i | sed "s/psmc//"))
	echo $id
	/opt/psmc/utils/psmc_plot.pl -u 1.6e-08 -g 5 ~/PSMC/plots/${id}plot $OUTPUT_path/$i
done
```

The different plots have been joined using AdobeIllustrator for better visualization of results and between species comparison.

## Bootstrapping

Next step will be to repeat the analysis with a Bootstrapping option to evaluate the confidence of the estimates made by PSMC.

## PseudoDiploidization of X chromosome

In order to infer population split times, a pseudodiploid individual made by taking 2 male X chromosome alignments joined together as if they were two alleles of a single individual. The time at which the Ne estimates goes to infinite, is the estimated time of split between the populations of the two individuals.

I'll first try with one pair of individuals:
c_lp_sm_0138 and c_ll_ki_0090
```
cd ~/PSMC
```

### 1. Get fq of X chromosome only

I use seqtk subseq:
```
seqtk subseq consensus_fq/c_lp_sm_0138.fq ~/CatGenome_CHR_BEDs/X_CHR_coordinates.bed \
> consensus_fq/c_lp_sm_0138.X_chr.fq

seqtk subseq consensus_fq/c_ll_ki_0090.fq ~/CatGenome_CHR_BEDs/X_CHR_coordinates.bed \
> consensus_fq/c_ll_ki_0090.X_chr.fq
```

### 2. Merge X_chr.fq of the two individuals

I use seqtk mergefa
```
seqtk mergefa consensus_fq/c_lp_sm_0138.X_chr.fq consensus_fq/c_ll_ki_0090.X_chr.fq \
> consensus_fq/c_lp_sm_0138-c_ll_ki_0090.X_chr.fq

# Got this message:
# [stk_mergefa] Different sequence names: X:1-130553893 != X:1-130553765
# [stk_mergefa] Unequal sequence length: 130553893 != 130553765
# [stk_mergefa] (same,diff,hom-het,het-hom,het-het)=(96408169,202582,25493,18903,4973)
# The file was generated so I will try running psmc with it to see if this is an error or normal seqtk mergefa behaviour
```

### 3. Convert my pseudohaploid sample fq to PSMC input format

I use fq2psmcfa
```
/opt/psmc/utils/fq2psmcfa consensus_fq/c_lp_sm_0138-c_ll_ki_0090.X_chr.fq \
> inputfiles_psmcfa/c_lp_sm_0138-c_ll_ki_0090.X_chr.psmcfa
```

### 4. Run PSMC on the pseudohaploid

```
psmc -p "4+25*2+4+6" -o output_psmc/c_lp_sm_0138-c_ll_ki_0090.X_chr.psmc \
inputfiles_psmcfa/c_lp_sm_0138-c_ll_ki_0090.X_chr.psmcfa
```

### 5. Paint results

```
/opt/psmc/utils/psmc_plot.pl -u 1.6e-08 -g 5 plots/c_lp_sm_0138-c_ll_ki_0090.X_chr.plot \
output_psmc/c_lp_sm_0138-c_ll_ki_0090.X_chr.psmc
```

As the code worked I wrote a version to loop through all useful combinations (no repetitions) of 5 individuals (1 per species + 2 for east and west lynx lynx). The script can be found at:
4.PSMC-executables/X_chr_PSMC_all.sh
```
screen -S X_chr_PSMC_all
script X_chr_PSMC_all.log
./X_chr_PSMC_all.sh
```
